File: ruby/lib/redis/commands/lrange.rb
require_relative './base'

module Redis
  module Commands
    class LRange < Base
      def self.command
        'lrange'
      end
      
      def run
        @mutex.synchronize do
          @key = args[0]
          @range = args[1..]
  
          return '-ERR wrong number of arguments for command' if invalid_command?
          return '-ERR value is not an integer or out of range' unless is_range_numeric?
          
          @range = @range.map(&:to_i)
          @store[@key] ||= []

          result = @store[@key][@range.first..@range.last]
          if result.empty?
            return Redis::Resp::Serializer.serialize('empty array')
          end

          Redis::Resp::Serializer.serialize(result)
        end  
      end

      private

      def invalid_command?
        return true if @key.empty? || @range.empty?
        return true if @range.size != 2  
        
        false
      end

      def is_range_numeric?
        valid = @range.map do |val|
          true if Integer(val)
          rescue
            false
        end

        !valid.any?(false)
      end
    end
  end
end

File: python/src/commands/lrange.py
from .base import BaseCommand
from src.resp.serializer import RespSerializer

class LRange(BaseCommand):
  @classmethod
  def command(cls) -> bytes:
    return b'lrange'
  
  def run(self) -> bytes:
    key: bytes = self.args[0]
    if len(self.args[1:]) != 2:
      return b'-ERR wrong number of arguments for command'
    
    if not isinstance(self.store[key], list):
      return b'-ERR Invalid datatype'
    
    start_range: int = int(self.args[1])
    end_range: int = int(self.args[2])
    if end_range == -1:
      return RespSerializer.serialize(data=self.store[key][start_range:])
    else:
      return RespSerializer.serialize(data=self.store[key][start_range:end_range + 1])
